plugins/statistics.koplugin/main.lua

```diff
@@ -2636,4 +2748,124 @@ function ReaderStatistics:getCurrentBookReadPages()
     return read_pages
 end
 
+function ReaderStatistics.onSync(local_path, cached_path, income_path)
+    local conn_income = SQ3.open(income_path)
+    local ok1, v1 = pcall(conn_income.rowexec, conn_income, "PRAGMA schema_version")
+    if not ok1 or tonumber(v1) == 0 then
+        -- no income db or wrong db, first time sync
+        logger.warn("statistics open income DB failed", v1)
+        return true
+    end
+
+    local sql = "attach '" .. income_path:gsub("'", "\\'") .."' as income_db;"
+    -- then we try to open cached db
+    local conn_cached = SQ3.open(cached_path)
+    local ok2, v2 = pcall(conn_cached.rowexec, conn_cached, "PRAGMA schema_version")
+    local attached_cache
+    if not ok2 or tonumber(v2) == 0 then
+        -- no cached or error, no item to delete
+        logger.warn("statistics open cached DB failed", v2)
+    else
+        attached_cache = true
+        sql = sql .. "attach '" .. cached_path:gsub("'", "\\'") ..[[' as cached_db;
+            -- first we delete from income_db books that exist in cached_db but not in local_db,
+            -- namely the ones that were deleted since last sync
```

Comment:
I'm having a hard time understanding how all this work.
You put some comments though, and there surely is a lot of thinking and cleverness in there - so may be it's just me, and me not being interested enough :)
An additional comment/section (in syncservice.lua I guess, as the working must be similar for stats & vocabs, and future usages) explaining how all this works might be welcome - some ASCII-art drawing of the flows too :)

You mentionned above that cached_db is "to be uploaded to server" - so is cached_db an adequate name?
Also, some comments about the lifetime of these files may help (cached_db stays?).

Reply:
A new paragraph is added to syncservice.lua explaining how it works. Hopefully it is clearer now.

Reply:
Thanks, that's indeed make the whole thing clearer :)
About the stats sync, one can delete page_stat_data for a book for one day (in CalendarView, tap a day, long press a book title). Will these deletion be synced ? (Just mentionning that - if not, up to you to ignore this edge feature :)

Reply:
Yes, these deletions will also be synced.
https://github.com/koreader/koreader/blob/0ee2dd92353a17892a87f3ea6db5b238be684784/plugins/statistics.koplugin/main.lua#L2833-L2845